pageserver/src/tenant.rs

```diff
@@ -1229,9 +1229,45 @@ impl Tenant {
 
     pub async fn apply_timeline_archival_config(
         &self,
-        _timeline_id: TimelineId,
-        _config: TimelineArchivalState,
+        timeline_id: TimelineId,
+        state: TimelineArchivalState,
     ) -> anyhow::Result<()> {
+        let timeline = self
+            .get_timeline(timeline_id, false)
+            .context("Cannot apply timeline archival config to inexistent timeline")?;
+
+        let is_archived = timeline.is_archived();
+
+        let upload_needed = match (is_archived, state) {
+            (Some(true), TimelineArchivalState::Archived)
+            | (Some(false), TimelineArchivalState::Unarchived) => {
+                // Nothing to do
+                tracing::info!("intended state matches present state");
+                None
+            }
+            (Some(false), TimelineArchivalState::Archived) => Some(Some(Utc::now().naive_utc())),
+            (Some(true), TimelineArchivalState::Unarchived) => Some(None),
+            (None, _) => bail!("timeline not ready yet, please retry"),
+        };
+
+        if let Some(intended_archived_at) = upload_needed {
+            // We have to schedule this into the upload queue with the other operations,
+            // as if we side-step the upload queue and do it immediately, later operations
+            // without the `archived_at` field might override the file.
+            // Deletion addresses this by shutting down the upload queue. We cannot afford
+            // this here.
+            timeline
+                .remote_client
+                .schedule_index_upload_for_archived_at_update(intended_archived_at)?;
+            const MAX_WAIT: Duration = Duration::from_secs(10);
+            let Ok(v) =
+                tokio::time::timeout(MAX_WAIT, timeline.remote_client.wait_completion()).await
+            else {
+                tracing::warn!("reached timeout for waiting on upload queue");
+                bail!("reached timeout for upload queue flush");
+            };
+            v?;
+        }
```

Comment:
I am just quickly looking from mobile, but I think there is a better design where we handle the needed update on remote timeline client. I am doing similar with the persistent gc blocking to make it retryable.

I intentionally did not allow inspection of the dirty indexpart outside of remote timeline client, and that is what I think should be done here as well.

Reply:
ok in the call I agreed that it makes sense to move it to the remote timeline client. offloaded timelines probably won't have a remote timeline client, but we'd need to duplicate the logic for them anyways, because for non-offloaded ones we have to use remote timeline client.